### PR TITLE
Add 'count' searchcache operator

### PR DESCRIPTION
**Pull request by lukebjerring**
## Description
Fixes https://github.com/web-platform-tests/wpt.fyi/issues/28

Adds support for a `count` operator in searchcache queries, where the number of runs in any given row that satisfy the requirement must be a precise number (count).

e.g. `count:3(status:pass|status:ok)`

Also added `three(...)`, `two(...)`, and `one(...)` aliases for simplicity.

### CODE REVIEW

**Comment by lukebjerring** [comment]
Example query:
[three(status:pass|status:ok)](https://count-dot-wptdashboard-staging.appspot.com/results/?label=master&product=chrome%5Bexperimental%5D&product=edge&product=firefox%5Bexperimental%5D&product=safari%5Bexperimental%5D&aligned&q=three%28status%3Apass%7Cstatus%3Aok%29)

File: api/query/atoms.go [hunk 1]
```diff
@@ -78,8 +78,10 @@ func (e AbstractExists) BindToRuns(runs ...shared.TestRun) ConcreteQuery {
 	queries := make([]ConcreteQuery, len(e.Args))
 	for i, arg := range e.Args {
 		var query ConcreteQuery
+		// For sequential + count, we pass all runs.
 		if _, isSeq := arg.(AbstractSequential); isSeq {
-			// For sequential, we pass all runs.
+			query = arg.BindToRuns(runs...)
+		} else if _, isCount := arg.(AbstractCount); isCount {
```
Comment (KyleJu) [inline api/query/atoms.go]: Looks like count operator is only implemented at the root level. If I understand correctly, a case like
 exists: [{
          and: [
            {pattern: 'a'},
           {
          count: 5,
          where: {
            or: [{status: 'PASS'}, {status: 'OK'}],
          },
        }
      ]}]

File: api/query/atoms.go [hunk 2]
```diff
@@ -529,6 +551,35 @@ func (e *AbstractSequential) UnmarshalJSON(b []byte) error {
 	return nil
 }
 
+// UnmarshalJSON for AbstractCount attempts to interpret a query atom as
+// {"count": int, "where": query}.
+func (c *AbstractCount) UnmarshalJSON(b []byte) error {
```
Comment (KyleJu) [inline api/query/atoms.go]: what is the reasoning behind BindToRuns implements on a AbstractCount while UnmarshalJSON implements on *AbstractCount?

File: api/query/atoms_test.go [hunk 3]
```diff
@@ -297,6 +297,34 @@ func TestStructuredQuery_sequential(t *testing.T) {
 			}}}, rq)
 }
 
+func TestStructuredQuery_count(t *testing.T) {
+	var rq RunQuery
+	err := json.Unmarshal([]byte(`{
+		"run_ids": [0, 1, 2],
+		"query": {
```
Comment (KyleJu) [inline api/query/atoms_test.go]: Question outside of this PR - if AbstractExists is the root, should every single query in this test file should start with "exists"?

File: webapp/components/test-search.js [hunk 4]
```diff
@@ -39,9 +39,20 @@ const QUERY_GRAMMAR = ohm.grammar(`
     Q = ListOf<RootExp, space*>
 
     RootExp
-      = "seq(" ListOf<Exp, space*> ")" -- sequential
+      = Sequential
+      | Count
```
Comment (KyleJu) [inline webapp/components/test-search.js]: count operator exists at the root level?

File: api/query/atoms.go [hunk 5]
```diff
@@ -575,5 +626,10 @@ func unmarshalQ(b []byte) (AbstractQuery, error) {
 	if err == nil {
 		return s, nil
 	}
+	var c AbstractCount
+	err = json.Unmarshal(b, &c)
+	if err == nil {
+		return c, nil
+	}
 	return nil, errors.New(`Failed to parse query fragment as test name pattern, test status constraint, negation, disjunction, or conjunction`)
```
Comment (KyleJu) [inline api/query/atoms.go]: nit: `Failed to parse query fragment as test name pattern, test status constraint, negation, disjunction, conjunction, sequential or count`

File: api/query/atoms.go (hunk 2 above)
Comment (lukebjerring) [reply api/query/atoms.go]: You pass a ref if A) you're going to mutate the ref or B) it's costly to copy (a large struct).

File: api/query/atoms_test.go (hunk 3 above)
Comment (lukebjerring) [reply api/query/atoms_test.go]: No; Server-side, AbstractExists is only the root when `{exists: ...}` is passed in from the client (which, it always is).

File: webapp/components/test-search.js (hunk 4 above)
Comment (lukebjerring) [reply webapp/components/test-search.js]: We shouldn't be combining it with other queries; you'd get unexpected behaviour

File: api/query/atoms.go (hunk 5 above)
Comment (lukebjerring) [reply api/query/atoms.go]: FYI the little +/- icon 
image

Lets you make a suggestion I can apply with the click of a button.

```suggestion
	return nil, errors.New(`Failed to parse query fragment as test name pattern, test status constraint, negation, disjunction, conjunction, sequential or count`)
```

**Pull request merged**